FEAT: PDF Converter Updates

KutalVolkan opened the pull request:
This PR continues the PDF Converter feature initiated by Pratyaksh0215 for the PyRIT project, addressing and resolving the issues highlighted in the original [PR #508](https://github.com/Azure/PyRIT/pull/508). 

**Supported Features**:
  1. **Templated PDF Generation**:
      - Supports YAML-based templates utilizing `SeedPrompt`.
      - Enables dynamic data injection into placeholders defined in the template.
  2. **Non-Templated PDF Generation**:
      - Directly converts plain string prompts into PDFs without any template.

#### **Related Issues**
- #508

jbolor21 commented on tests/converter/test_pdf_converter.py:
```diff
@@ -0,0 +1,136 @@
+    """A PDFConverter with a valid template file."""
+    template_content = "This is a test template. Prompt: {{ prompt }}"
+    template_file = tmp_path / "test_template.txt"
+    template_file.write_text(template_content)
```
nit: I think this creates a test template file but should this file be cleaned up and removed after the test completes successfully?

rdheekonda commented on .idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
It seems the `.idea` folder has been added to the repository. Since `.idea` contains IDE-specific settings, it is generally not needed in version control unless explicitly required for team-wide configurations. 

Could you consider:
1. Adding `.idea/` to the .gitignore file to prevent it from being tracked in the future.
2. Remove the `.idea` folder from the repository 

This will help keep the repository clean and avoid including IDE-specific settings that may not be relevant to all contributors.

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -25,6 +26,7 @@ def data_serializer_factory(
     data_type: PromptDataType,
     value: Optional[str] = None,
     extension: Optional[str] = None,
+    memory: Optional[MemoryInterface] = None,
```
Since memory is configured centrally, so we don't need to pass it to individual components. I suggest removing it here, similar to how it was handled for ImageDataTypeSerializer and other components

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -36,7 +38,7 @@ def data_serializer_factory(
             return ErrorDataTypeSerializer(prompt_text=value)
         elif data_type == "url":
-            return URLDataTypeSerializer(prompt_text=value)
+            return URLDataTypeSerializer(prompt_text=value, memory=memory)
```
Consider removing memory argument here

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -59,16 +61,16 @@ class DataTypeSerializer(abc.ABC):
-    @property
-    def _memory(self) -> MemoryInterface:
+    def __init__(self, memory: Optional[MemoryInterface] = None):
+        # Local import to prevent circular dependency
         from pyrit.memory import CentralMemory
 
-        return CentralMemory.get_memory_instance()
+        self._memory = memory or CentralMemory.get_memory_instance()
```
Since memory is configured centrally, so we don't need to pass it to individual components. I suggest removing it here, similar to how it was handled for ImageDataTypeSerializer and other components

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -216,7 +218,8 @@ def _is_azure_storage_url(self, path: str) -> bool:
 
 class TextDataTypeSerializer(DataTypeSerializer):
-    def __init__(self, *, prompt_text: str):
+    def __init__(self, *, prompt_text: str, memory: Optional[MemoryInterface] = None):
+        super().__init__(memory=memory)
```
We don't need memory to be passed to the Serializers, we have configured it centrally. So could you please revert the changes.

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -225,7 +228,8 @@ def data_on_disk(self) -> bool:
 class ErrorDataTypeSerializer(DataTypeSerializer):
-    def __init__(self, *, prompt_text: str):
+    def __init__(self, *, prompt_text: str, memory: Optional[MemoryInterface] = None):
+        super().__init__(memory=memory)
```
Same comment as above

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -234,16 +238,26 @@ def data_on_disk(self) -> bool:
 class URLDataTypeSerializer(DataTypeSerializer):
-    def __init__(self, *, prompt_text: str):
+    def __init__(self, *, prompt_text: str, memory: Optional[MemoryInterface] = None):
+        super().__init__(memory=memory)
```
Same comment as above. We don't need memory here.

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -234,16 +238,26 @@ def data_on_disk(self) -> bool:
+        *,
+        prompt_text: Optional[str] = None,
+        extension: Optional[str] = None,
+        memory: Optional[MemoryInterface] = None,
```
Same comment as above, no need to pass memory here.

rdheekonda commented on pyrit/models/data_type_serializer.py:
```diff
@@ -261,7 +275,9 @@ def __init__(
         extension: Optional[str] = None,
+        memory: Optional[MemoryInterface] = None,
     ):
+        super().__init__(memory=memory)
```
Could we revert the changes, no need to pass memory.

rdheekonda commented on pyrit/prompt_converter/pdf_converter.py:
```diff
@@ -0,0 +1,104 @@
+        page_height: Optional[int] = 297,
+    ) -> None:
+        self._template_path = template_path
+        self._memory = memory
```
Passing memory is not required. Could you use [this ](https://github.com/Azure/PyRIT/blob/4d38c73259b4815d0dcb84521410f264619906ee/pyrit/prompt_converter/add_text_image_converter.py#L121C9-L136C60) example as a reference for saving the PDF to disk? It abstracts saving to the local disk versus Azure Storage based on the configured memory.

rdheekonda commented on pyrit/prompt_converter/pdf_converter.py:
```diff
@@ -0,0 +1,104 @@
+        # Prepare the content
+        if self._template_path:
+            try:
+                with open(self._template_path, "r") as file:
```
Do we have a sample template file? If not, I suggest adding one to this PR to help users understand

rdheekonda commented on tests/converter/test_pdf_converter.py:
```diff
@@ -0,0 +1,136 @@
+
+
+@pytest.fixture
+def mock_memory():
+    """Create a mock MemoryInterface using MagicMock."""
+    memory_mock = MagicMock(spec=MemoryInterface)
+
+    # Mock necessary attributes used by PDFConverter
+    memory_mock.results_storage_io = MagicMock()
+    memory_mock.results_path = "/mock/results/path"
+
+    # Mock all abstract methods with default return values
+    memory_mock._add_embeddings_to_memory.return_value = None
+    memory_mock._get_prompt_pieces_by_orchestrator.return_value = []
+    memory_mock._get_prompt_pieces_with_conversation_id.return_value = []
+    memory_mock._init_storage_io.return_value = None
+    memory_mock.add_request_pieces_to_memory.return_value = None
+    memory_mock.dispose_engine.return_value = None
+    memory_mock.get_all_embeddings.return_value = []
+    memory_mock.get_all_prompt_pieces.return_value = []
+    memory_mock.get_prompt_request_pieces_by_id.return_value = []
+    memory_mock.insert_entries.return_value = None
+    memory_mock.insert_entry.return_value = None
+    memory_mock.query_entries.return_value = []
+    memory_mock.update_entries.return_value = None
+
+    return memory_mock
```
For mocking memory, I would suggest following the [reference ](https://github.com/Azure/PyRIT/blob/4d38c73259b4815d0dcb84521410f264619906ee/tests/unit/converter/test_add_text_image_converter.py#L19C1-L22C1)here and remove this?

rdheekonda commented on tests/converter/test_pdf_converter.py:
```diff
@@ -0,0 +1,136 @@
+        result = await converter.convert_async(prompt=prompt)
+        assert isinstance(result, ConverterResult)
+        assert result.output_text == "mock_url"
+        serializer_mock.save_data.assert_called_once()
```
Since this is a text-to-PDF converter, could you consider adding a real end-to-end test without mocking since it doesn't require any authentication? This would involve creating a PDF, verifying its content, and then deleting the file after the test.

KutalVolkan replied:
In the current implementation, we:
- Generate a PDF using the provided prompt.
- Save the generated PDF to a temporary file.
- Verify that the file exists and is not empty.
- Clean up by deleting the file after the test.

However, verifying the actual written content of the PDF is not easily feasible without adding another dependency for PDF parsing (e.g., `pypdf 5.1.0`). Without such a dependency, validating the PDF's content would require manual parsing of the binary file, which is error-prone.

Here's the current test:

```python
@pytest.mark.asyncio
async def test_convert_async_end_to_end_no_reader(tmp_path):
    prompt = "Test for PDF generation."
    pdf_file_path = tmp_path / "output.pdf"
    converter = PDFConverter(template_path=None)

    result = await converter.convert_async(prompt=prompt)
    with open(pdf_file_path, "wb") as pdf_file:
        pdf_file.write(result.output_text.encode("latin1"))

    assert pdf_file_path.exists()
    assert os.path.getsize(pdf_file_path) > 0
    pdf_file_path.unlink()
``` 

If you have other ideas or if we are allowed to use additional dependencies, please let me know.

rdheekonda replied:
Testing the actual content is important, but it may not be necessary for this PR since we've started adding integration tests to the repository. As such, this function is sufficient for now. Thanks for looking into alternatives!

KutalVolkan commented:
Hello @rdheekonda & @jbolor21,
 
The current implementation saves the PDF output with a `.txt` extension because the `URLDataTypeSerializer` is hardcoded to use `txt` as the file extension. To fix this, I’m thinking of introducing a `PDFDataTypeSerializer`, like this:

```python
class PDFDataTypeSerializer(DataTypeSerializer):
    def __init__(self, *, prompt_text: str):
        self.data_type = "pdf"
        self.value = prompt_text
        self.data_sub_directory = "/dbdata/pdfs"  # Update to reflect PDF storage
        self.file_extension = "pdf"

    def data_on_disk(self) -> bool:
        return True
```

Does this approach make sense? My brain feels like it’s in low-power mode, so I’d appreciate a sanity check! ☕️🤖

rdheekonda commented:
> Hello @rdheekonda & @jbolor21,
> 
> The current implementation saves the PDF output with a `.txt` extension because the `URLDataTypeSerializer` is hardcoded to use `txt` as the file extension. To fix this, I’m thinking of introducing a `PDFDataTypeSerializer`, like this:
> 
> ```python
> class PDFDataTypeSerializer(DataTypeSerializer):
>     def __init__(self, *, prompt_text: str):
>         self.data_type = "pdf"
>         self.value = prompt_text
>         self.data_sub_directory = "/dbdata/pdfs"  # Update to reflect PDF storage
>         self.file_extension = "pdf"
> 
>     def data_on_disk(self) -> bool:
>         return True
> ```
> 
> Does this approach make sense? My brain feels like it’s in low-power mode, so I’d appreciate a sanity check! ☕️🤖

We don't need a dedicated `PDFDataTypeSerializer` for handling PDFs. Instead, we can leverage the `URLDataTypeSerializer` by passing the file extension when calling `data_type_serializer` method. In addition, we need to update the factory method with a condition like:

```python
elif data_type == "url":
    return UrlDataTypeSerializer(prompt_text=value, extension=extension)
```

This approach generalizes the serializer to handle various extensions, such as .txt, .xls, etc., making it more flexible and reusable.

KutalVolkan commented:
Hello @rdheekonda,

Feel free to review. I will go over it tomorrow with a fresh mind. I will be happy to clean up and resolve any upcoming issues :)

rdheekonda commented on pyrit/prompt_converter/pdf_converter.py:
```diff
@@ -0,0 +1,145 @@
+        pdf = FPDF(format=(self._page_width, self._page_height))  # Use custom page size
+        pdf.add_page()
+        pdf.set_font(self._font_type, size=self._font_size)  # Use custom font settings
+        pdf.multi_cell(0, 10, content)
```
Consider creating instance variables for first 2 args of multi_cell maybe column_width and row_height so users can pass those if they would like

rdheekonda commented on pyrit/datasets/prompt_templates/pdf/red_teaming_application_template.txt:
```diff
@@ -0,0 +1,11 @@
+Dear {{ hiring_manager_name }},
```
We already have a `prompt_converter` directory under datasets. I suggest creating a `pdf_converters` directory and moving these two yaml files there. 

Additionally, I recommend updating these templates format to align with the example converter template located at `pyrit\datasets\prompt_templates\math_prompt_template.yaml`, ensuring compatibility with the `SeedPrompt` object.

rdheekonda commented on doc/code/converters/pdf_converter.py:
```diff
@@ -0,0 +1,67 @@
+from pyrit.common.path import DATASETS_PATH
+
+# %% Templates and Dynamic Data
+template_folder = pathlib.Path(DATASETS_PATH) / "prompt_templates" / "pdf"
```
After refactoring the YAML template, consider testing it in the same way as other converter demo notebooks. For example, use the approach: prompt_template = SeedPrompt.from_yaml_file(prompt_template_path). Ref: `doc\code\converters\math_prompt_converter`

rdheekonda reviewed (approved):
Thanks for addressing all the comments. It looks great! Approved with a few minor comments—once those are addressed, it's good to merge.

KutalVolkan commented:
> Thanks for addressing all the comments. It looks great! Approved with a few minor comments—once those are addressed, it's good to merge.

Hello @rdheekonda,

It would be great if you could take another look. Please let me know if anything goes against your feedback, and I’ll be ready to resolve it. :)
